File: leet-code/sumNumbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * 递归，深度优先遍历
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function (root) {
  let sum = 0
  let dfs = function (curr, str = '') {
    if (curr == null) {
      return 0
    }
    if (curr.left == null && curr.right == null) {
      return (sum += Number(str + curr.val))
    }
    dfs(curr.left, str + curr.val)
    dfs(curr.right, str + curr.val)
  }
  dfs(root)
  return sum
}

console.log(sumNumbers({ val: 1, left: { val: 2 }, right: { val: 3 } }))

/**
 * 队列，广度优先遍历
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers2 = function (root) {
  let queueNode = []
  let queueVal = []
  let sum = 0

  queueNode.push(root)
  queueVal.push(root.val)

  while (queueNode.length != 0) {
    let tempNode = queueNode.shift()
    let tempVal = queueVal.shift()
    console.log(tempNode, tempVal)

    if (tempNode.left == null && tempNode.right == null) {
      sum += tempVal
      return
    }

    if (tempNode.left) {
      queueNode.push(tempNode.left)
      queueVal.push(Number(tempVal + tempNode.left.val))
    }
    if (tempNode.right) {
      queueNode.push(tempNode.right)
      queueVal.push(Number(tempVal + tempNode.right.val))
    }
  }
  return sum
}
console.log(sumNumbers2({ val: 1, left: { val: 2 }, right: { val: 3 } }))

File: leet-code/sumNumbers.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// 递归，深度优先遍历
func sumNumbers(root *TreeNode) int {
	sum := 0
	var dfs func(root *TreeNode, val int)
	dfs = func(root *TreeNode, val int) {
		if root == nil {
			return
		}
		if root.Left == nil && root.Right == nil {
			sum += val*10 + root.Val
		}
		dfs(root.Left, val*10+root.Val)
		dfs(root.Right, val*10+root.Val)
	}
	dfs(root, 0)
	return sum
}

// 队列，广度优先遍历
func sumNumbers2(root *TreeNode) int {
	sum := 0
	if root == nil {
		return sum
	}
	queueNode := make([]*TreeNode, 0)
	queueVal := make([]int, 0)
	// 初始化
	queueNode = append(queueNode, root)
	queueVal = append(queueVal, root.Val)

	for len(queueNode) > 0 {
		// 出队
		tempNode := queueNode[0]
		queueNode = queueNode[1:]
		tempVal := queueVal[0]
		queueVal = queueVal[1:]

		if tempNode.Left == nil && tempNode.Right == nil {
			sum += tempVal
		}

		if tempNode.Left != nil {
			queueNode = append(queueNode, tempNode.Left)
			queueVal = append(queueVal, tempVal*10+tempNode.Left.Val)
		}
		if tempNode.Right != nil {
			queueNode = append(queueNode, tempNode.Right)
			queueVal = append(queueVal, tempVal*10+tempNode.Right.Val)
		}
	}

	return sum
}
